### PR TITLE
Deploy linking in pypechain. Includes version bump to v0.0.37

### LINKED ISSUE #118: [ISSUE] Support for contract linking in pypechain
Currently, pypechain doesn't support linking contracts in the bytecode exposed in generated types. We work around this in `agent0` by doing a string replace on the bytecode itself:

https://github.com/delvtech/agent0/blob/1643282d986f6dbf1bad37c1f6f496d60b76056f/src/agent0/ethpy/hyperdrive/deploy.py#L561
https://github.com/delvtech/agent0/blob/1643282d986f6dbf1bad37c1f6f496d60b76056f/src/agent0/ethpy/hyperdrive/deploy.py#L412

Ideally, pypechain would allow the `deploy` function (and anything else requiring the bytecode) to accept the linked contract address.

### PR DESCRIPTION
**Pull request by slundqui**
https://github.com/delvtech/pypechain/issues/118

This PR handles the case where bytecode requires a linked reference to a previously deployed contract. Here, the python generated types add in a required `link_references` argument to `deploy` if this is found in the abi.

- Support for link references in pypechain generated files.
- Using forge instead of solc in tests. Ideally we would run tests with supported solidity compilers.
- Consolidating tests readme to outermost readme.
- Renamed `constructorArgs` to `constructor_args` in pypechain generated types.

### CODE REVIEW

File: pypechain/test/deploy_linking/types/ContractContract.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+        except FallbackNotFound:
+            print("Fallback function not found. Continuing...")
+
+    functions: ContractContractFunctions
+
+    class LinkReferences(NamedTuple):
+        """Link references required when deploying."""
+
+        my_library: MyLibraryContract
```
Comment (slundqui) [inline pypechain/test/deploy_linking/types/ContractContract.py:139]: @sentilesdal do you have thoughts on this? I was going back and forth between the pypechain contract object and an address here. On one hand, requiring the object gives strong type guarantees, but require the linked contract to be built by pypechain to use, whereas giving an address is more flexible.

File: pypechain/test/deploy_linking/test_deployment.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        # This should fail: Contract requires linked references
+        contract = ContractContract.deploy(w3, local_account)
+
+    def test_deployment_with_contract_linking(self, w3: Web3, local_account: LocalAccount):
+        """Tests deployment using contract linking in deploy."""
+
+        # Deploy the library first
+        library_contract = MyLibraryContract.deploy(w3, local_account)
+
+        #
```
Comment (sentilesdal) [inline pypechain/test/deploy_linking/test_deployment.py]: empty comment

File: pypechain/test/deploy_linking/test_deployment.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
+"""Tests for overloading methods."""
```
Comment (sentilesdal) [inline pypechain/test/deploy_linking/test_deployment.py]: 'Tests for deploying contracts with linked libraries. '
Rename this file to test_deploy_linking.py

File: pypechain/test/deploy_linking/test_deployment.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        """Tests deployment using contract linking in deploy."""
+
+        # Deploy the library first
+        library_contract = MyLibraryContract.deploy(w3, local_account)
+
+        #
+        contract = ContractContract.deploy(
+            w3,
+            local_account,
+            link_references=ContractContract.LinkedReferences(
```
Comment (sentilesdal) [inline pypechain/test/deploy_linking/test_deployment.py]: 👏

File: pypechain/foundry/utilities.py [hunk 5]
```diff
@@ -14,3 +17,23 @@ def is_foundry_json(val: object) -> TypeGuard[FoundryJson]:
...
+    # Rework link references to be a dictionary keyed by a contract name, valued with
+    # the hex code placeholder in the bytecode.
+    out_link_references = []
+    for function_path, value in link_references.items():
+        for contract_name, _ in value.items():
+            fully_qualified_lib_name = f"{function_path}:{contract_name}"
+            hex_code = Web3.keccak(text=fully_qualified_lib_name).hex()
+            # Remove the leading 0x, and take the 34 character prefix, and add the placeholder
+            # token
+            placehoder_code = f"__{hex_code[2:36]}__"
```
Comment (sentilesdal) [inline pypechain/foundry/utilities.py]: you need to wrap with `__$...$__`

File: pypechain/utilities/abi.py [hunk 6]
```diff
@@ -672,14 +675,23 @@ def load_abi_infos_from_file(file_path: Path) -> list[AbiInfo]:
             bytecode = get_bytecode_from_json(json_file)
             # hardhat jsons have the contract name as a field.
             contract_name = json_file.get("contractName")
-            return [AbiInfo(abi=abi, bytecode=bytecode, contract_name=contract_name)]
+            # TODO
```
Comment (sentilesdal) [inline pypechain/utilities/abi.py:678]: whatcha wanna TODO?

Comment (sentilesdal) [reply pypechain/utilities/abi.py:678]: add TODOs to get link references for hardhat and solc as well.  should just work though i think

File: pypechain/utilities/bytecode.py [hunk 7]
```diff
@@ -19,3 +25,17 @@ def get_bytecode_from_json(json_abi: HardhatJson | FoundryJson) -> str:
         return get_bytecode_from_hardhat_json(json_abi)
 
     raise ValueError("Unable to retrieve bytecode, JSON in unknown format.")
+
+
+def get_bytecode_link_references_from_json(json_abi: HardhatJson | FoundryJson) -> list[LinkReferences]:
```
Comment (sentilesdal) [inline pypechain/utilities/bytecode.py:30]: add SolcJson?

File: pypechain/render/contract.py [hunk 8]
```diff
@@ -488,3 +497,42 @@ def get_event_datas(abi: ABI) -> dict[str, EventData]:
...
+        A data structure containing the link reference data
+    """
+    contract_name_keys = []
+    contract_name_key_and_types = []
+    contract_types = []
+    contract_name_key_to_placeholder_lookup = []
+    for link in link_references:
+        contract_name = link["contract_name"]
+        # The contract_name_key is reformatting the contract name to be snake case
+        contract_name_key = camel_to_snake(contract_name)
```
Comment (sentilesdal) [inline pypechain/render/contract.py:522]: maybe a nit here, but I like when we preserve names of files and contracts 'as-is'.  it makes it more clear what we're trying to match when we provide addresses.

File: pypechain/templates/contract.py/contract.py.jinja2 [hunk 9]
```diff
@@ -2,9 +2,13 @@ class {{contract_name}}Contract(Contract):
     """A web3.py Contract class for the {{contract_name}} contract."""
 
     abi: ABI = {{contract_name | lower}}_abi
-    {%- if has_bytecode %}
-    bytecode: bytes | None = try_bytecode_hexbytes({{contract_name | lower}}_bytecode, "{{contract_name | lower}}")
-    {%- endif %}
+    # We change `bytecode` as needed for linking, but keep
+    # `_raw_bytecode` unchanged as an original copy.
```
Comment (sentilesdal) [inline pypechain/templates/contract.py/contract.py.jinja2:6]: nice

File: pypechain/utilities/format.py [hunk 10]
```diff
@@ -110,6 +111,32 @@ def capitalize_first_letter_only(string: str) -> str:
...
+    ---------
+    camel_string: str
+        The string to convert.
+
+    Returns
+    -------
+    str
+        The snake case string.
+    """
+    return re.sub("((?<=[a-z0-9])[A-Z]|(?!^)[A-Z](?=[a-z]))", r"_\1", camel_string).lower()
```
Comment (sentilesdal) [inline pypechain/utilities/format.py:137]: lol

File: pypechain/test/deploy_linking/types/ContractContract.py (hunk 1 above)
Comment (sentilesdal) [reply pypechain/test/deploy_linking/types/ContractContract.py:139]: I can imagine a world where you might just grab the abi for the contract you're working with that already comes with a link reference... so address support would be really nice.  but in the world of pypechain its actually a little hard to imagine that you'd be compiling a contract without the link reference's source code.  let's leave with contracts for now and add address support later when that is necessary.

File: pypechain/test/deploy_linking/types/ContractContract.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+
+    link_references_placeholder_lookup: dict[str, str] = {"my_library": "__$81c732ea87169659ae18eec7be97daeb59$__"}
```
Comment (sentilesdal) [inline pypechain/test/deploy_linking/types/ContractContract.py:141]: looking at this i do feel more strongly about leaving the keys as capital case or whatever they were where they came from.

Comment (sentilesdal) [reply pypechain/test/deploy_linking/types/ContractContract.py:141]: one arguments is that it makes for easier grepping, another argument is what if it is something weird like "MY_wEiRd_cOnTrAct"

**Review by sentilesdal — APPROVED** [review]
great work, one thing about the keys and then a couple comments

File: pypechain/foundry/utilities.py (hunk 5 above)
Comment (sentilesdal) [reply pypechain/foundry/utilities.py]: nvm, saw u fixed this

**Pull request merged**